### DISCUSSION TITLE
Persistent marker

### DISCUSSION THREAD

**Discussion by cristianboicu01** [patrykandpatrick/vico#859] [Questions]
### Question

I would like to have a single marker, which I can drag along the line chart and it will not disappear after I release it. How can I achieve this kind of behaviour?

The only problem right now is that I cannot figure out a way on how to keep the marker visible on the last point it was dragged to.

Thank you!

### Vico version(s)

last

### UI framework(s)

Jetpack Compose

**Comment by Tyler-Lopez** [comment]
The way I have achieved this is through an invisible `marker` (because you otherwise frustratingly don't get `markerVisibilityListener` updates) and then listening for changes to `markerVisibilityListener` and using a `persistentMarker` off of that changed value.

**Reply by cristianboicu01** [reply]
Thank you for your suggestion. I tried something similar before, but I can't figure out where do I get the new position of the marker from that listener. 

I tried this approach, though I'm not sure if it's the right way
` override fun onShown(marker: CartesianMarker, targets: List<CartesianMarker.Target>) {
          persistentMarkerIndex = targets.first().x.toInt()
        }`
 
 Also, didn't yet figured out why but ` persistentMarkers = { marker at data[persistentMarkerIndex] }` fails to update marker position
 
 ### **Edit:** 
 
 Figured out I can get the last marker position by overriding `onUpdated` method. But still cant manage to update persistent marker position though I tried to use `rememberExtraLambda` with the index as a key.

**Comment by cristianboicu01** [comment]
I figured out why we can't update the `persistentMarker` position. 

The problem is that `persistentMarkers` remains the same object, and its hash code is the same, therefore the code in the if block executes only the first time.

Is this a bug or, am I doing something wrong?

    val persistentMarkerHashCode = Objects.hash(persistentMarkers, model.extraStore)
    if (persistentMarkerHashCode != previousPersistentMarkerHashCode) {
      persistentMarkerMap.clear()
      persistentMarkers?.invoke(persistentMarkerScope, model.extraStore)
      previousPersistentMarkerHashCode = persistentMarkerHashCode
    }

**Reply by Praveen-Android** [reply]
@cristianboicu01 were you able to get this to work?. If so, could you please share some details. I currently am looking to solve a similar use case. Thanks!

**Reply by cristianboicu01** [reply]
That's how I solved it. 
```
  var persistentMarkerIndex by remember { mutableIntStateOf(data.lastIndex) }

  val persistentMarkers = remember {
    mutableStateOf<(CartesianChart.PersistentMarkerScope.(ExtraStore) -> Unit)>(
      {
        marker at persistentMarkerIndex
      },
    )
  }
```

**Comment by Gowsky** [comment]
Hello! As such, this is a feature that isn't yet available. It was requested in #356. Once it's been introduced, there will be a dedicated API that helps achieve this behavior easily. In the meantime, you can use `persistentMarkers`-based workarounds or, for greater robustness, override `CartesianChart.drawMarker` as required.

> The way I have achieved this is through an invisible `marker` (because you otherwise frustratingly don't get `markerVisibilityListener` updates) and then listening for changes to `markerVisibilityListener` and using a `persistentMarker` off of that changed value.

This workaround sounds good, @Tyler-Lopez.

> I tried this approach, though I'm not sure if it's the right way
> `override fun onShown(marker: CartesianMarker, targets: List<CartesianMarker.Target>) { persistentMarkerIndex = targets.first().x.toInt() }`

Yes, this works, @cristianboicu01.

> The problem is that `persistentMarkers` remains the same object, and its hash code is the same, therefore the code in the if block executes only the first time.
> 
> Is this a bug or, am I doing something wrong?

The latter. You should use something like this:

```kt
persistentMarkers = rememberExtraLambda(marker, persistentMarkerIndex) { marker at persistentMarkerIndex }
```

It will then update as expected.

> That's how I solved it. 
> ```
>   var persistentMarkerIndex by remember { mutableIntStateOf(data.lastIndex) }
> 
>   val persistentMarkers = remember {
>     mutableStateOf<(CartesianChart.PersistentMarkerScope.(ExtraStore) -> Unit)>(
>       {
>         marker at persistentMarkerIndex
>       },
>     )
>   }
>```

This works, but the approach I suggested above is preferable.

---

Updates regarding the introduction of this feature will be shared under #356.

**Note that there's a recomposition issue related to strong skipping mode in version `2.0.0-beta.1` (#900)**, and it may get in the way of `persistentMarkers`-based workarounds. It will be fixed in the next update.

@Praveen-Android, you asked for our input here a while ago, so I wanted to tag you.

**Comment by Lynxfox** [comment]
@Gowsky is it possible to get a rough timeline for possible release of this dedicated API for markers? It would just help to know if we can wait for this or if we should dedicate effort in workarounds.

**Reply by Gowsky** [reply]
Hello, @Lynxfox. Because this is a free and open-source project we develop in our spare time, we can't provide any timelines. However, we're open to pull requests.
